ext/thymeleaf/src/main/java/org/mvcspec/ozark/ext/thymeleaf/CDIWebContext.java

```diff
@@ -0,0 +1,145 @@
+ *     <li>variables from the model</li>
+ *     <li>named beans from the cdi context</li>
+ * </ol>
+ * </p>
+ *
+ * @author Gregor Tudan
+ */
+class CDIWebContext implements IWebContext {
+
+    private final HttpServletRequest request;
+    private final HttpServletResponse response;
+    private final ServletContext context;
+    private final Locale locale;
+
+    private final BeanManager beanManager;
+    private final Set<String> cdiNamedBeans;
+
+    private final Map<String, Object> variables = new HashMap<>();
+
+    CDIWebContext(BeanManager beanManager, HttpServletRequest request, HttpServletResponse response, ServletContext servletContext, Locale locale) {
+        this.beanManager = beanManager;
+        this.request = request;
+        this.response = response;
+        this.context = servletContext;
+        this.locale = locale;
+        this.cdiNamedBeans = enumerateNamedBeans();
+    }
+
+    @SuppressWarnings("serial")
+    private Set<String> enumerateNamedBeans() {
+        Set<Bean<?>> beans = beanManager.getBeans(Object.class, new AnnotationLiteral<Any>() { });
+        return beans.stream().map(Bean::getName).filter(Objects::nonNull).collect(Collectors.toSet());
+    }
+
+    @Override
+    public HttpServletRequest getRequest() {
+        return request;
+    }
+
+    @Override
+    public HttpServletResponse getResponse() {
+        return response;
+    }
+
+    @Override
+    public HttpSession getSession() {
+        return request.getSession();
+    }
+
+    @Override
+    public ServletContext getServletContext() {
+        return context;
+    }
+
+    @Override
+    public Locale getLocale() {
+        return locale;
+    }
+
+    @Override
+    public boolean containsVariable(String name) {
+        Objects.requireNonNull(name, "The variable name must not be null");
+        if (variables.containsKey(name)) {
+            return true;
+        }
+        return cdiNamedBeans.contains(name);
+    }
+
+    @Override
+    public Set<String> getVariableNames() {
+        final Set<String> variableNames = new HashSet<>(variables.keySet());
+        variableNames.addAll(this.cdiNamedBeans);
+        return variableNames;
+    }
+
+    @Override
+    public Object getVariable(String name) {
+        Objects.requireNonNull(name, "The variable name must not be null");
+        if (variables.containsKey(name)) {
+            return variables.get(name);
+        } else if (cdiNamedBeans.contains(name)) {
+            return getCdiBean(name);
+        } else {
+            return null;
+        }
+    }
+
+    private Object getCdiBean(String name) {
+        return new LazyContextVariable<Object>() {
+            @Override
+            protected Object loadValue() {
+                Bean<?> bean = beanManager.getBeans(name).iterator().next();
+                CreationalContext ctx = beanManager.createCreationalContext(bean);
+                return beanManager.getReference(bean, bean.getBeanClass(), ctx);
```

Comment:
To be honest, I'm not sure if we create a memory leak here. I ran into a similar issue with one of our client projects recently where we used `Instance<SomeBean>` for lazily resolving CDI beans which weren't release correctly.

I guess this can only happen for `@Dependent` scoped beans, but as this is the default if you forget about the annotation, it may be dangerous. Maybe we can find a way keep track of the `CreationalContext` instances and then call `CreationalContext#release()` to destroy the dependent-scoped instances?

Or maybe I'm wrong and there won't be a leak? Not completely sure. It is still too early in the morning. ;-)

Reply:
Quiet possible - I need to do some reading on this. I remember seeing memory leaks on non-contextual objects (when passing null as creational context), but in this case I would expect CDI to handle the beans as defined by their context.

The thymeleaf-context does not have any lifecycle hooks, so we might need to find another way of knowing when the request has finished.

Reply:
> but in this case I would expect CDI to handle the beans as defined by their context.

Sure, but if there isn't a scope annotation it will be `@Dependent` scoped and therefore depend on the lifecycle of the class of the injection point. But in our case there isn't an injection point as we are doing the lookup manually. But as I said, I'm also not 100% sure about this.

> The thymeleaf-context does not have any lifecycle hooks, so we might need to find another way of knowing when the request has finished.

We could simply define a `close()` method on our context class and call it after the template has been evaluated in a `finally` block. I guess this should work.

Reply:
Ah, I get it now  - you are right, the dependent scope probably won't work and most things I read point out that we'll have to take care of these beans by disposing the creational context.

I'll take a shot at it tomorrow.

Reply:
Awesome! Thanks a lot! :beers:

Reply:
It was easier than I thought, seems to work fine.
I also noticed that thymeleaf does not flush buffers after rendering, so I did that before dropping into the disposing to improve the initial site response. The user will get the site rendered but the loading indicator will keep going a little longer in case a bean takes longer to dispose.